tests/suite/graceful_recovery_test.go

```diff
@@ -0,0 +1,316 @@
+	ngfContainerName   = "nginx-gateway"
+)
+
+var _ = Describe("Graceful Recovery test", Ordered, Label("nfr", "graceful-recovery"), func() {
+	files := []string{
+		"graceful-recovery/cafe.yaml",
+		"graceful-recovery/cafe-secret.yaml",
+		"graceful-recovery/gateway.yaml",
+		"graceful-recovery/cafe-routes.yaml",
+	}
+	ns := &core.Namespace{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: "graceful-recovery",
+		},
+	}
+
+	BeforeEach(func() {
+		Expect(resourceManager.Apply([]client.Object{ns})).To(Succeed())
+		Expect(resourceManager.ApplyFromFiles(files, ns.Name)).To(Succeed())
+		Expect(resourceManager.WaitForAppsToBeReady(ns.Name)).To(Succeed())
+
+		Expect(waitForWorkingTraffic()).ToNot(HaveOccurred())
+	})
+
+	AfterAll(func() {
+		Expect(resourceManager.DeleteFromFiles(files, ns.Name)).To(Succeed())
+		Expect(resourceManager.Delete([]client.Object{ns})).To(Succeed())
+	})
+
+	It("recovers when NGF container is restarted", func() {
+		runRecoveryTest(ngfContainerName, files, ns)
+	})
+
+	It("recovers when nginx container is restarted", func() {
+		runRecoveryTest(nginxContainerName, files, ns)
+	})
+})
+
+func runRecoveryTest(containerName string, files []string, ns *core.Namespace) {
+	podNames, err := framework.GetReadyNGFPodNames(k8sClient, ngfNamespace, releaseName, timeoutConfig.GetTimeout)
+	Expect(err).ToNot(HaveOccurred())
+	Expect(podNames).ToNot(BeEmpty())
+
+	leaseName, err := getLeaderElectionLeaseHolderName()
+	Expect(err).ToNot(HaveOccurred())
+
+	restartContainer(containerName)
+
+	checkContainerLogsForErrors(podNames[0])
+
+	Expect(waitForLeaderLeaseToChange(leaseName)).ToNot(HaveOccurred())
+
+	Expect(waitForWorkingTraffic()).ToNot(HaveOccurred())
+
+	Expect(resourceManager.DeleteFromFiles(files, ns.Name)).To(Succeed())
+
+	Expect(waitForFailingTraffic()).ToNot(HaveOccurred())
+
+	Expect(resourceManager.ApplyFromFiles(files, ns.Name)).To(Succeed())
+	Expect(resourceManager.WaitForAppsToBeReady(ns.Name)).To(Succeed())
+
+	Expect(waitForWorkingTraffic()).ToNot(HaveOccurred())
+}
+
+func restartContainer(containerName string) {
+	var jobScript string
+	if containerName == "nginx" {
+		jobScript = "PID=$(pgrep -f \"[n]ginx: master process\") && kill -9 $PID"
+	} else {
+		jobScript = "PID=$(pgrep -f \"/[u]sr/bin/gateway\") && kill -9 $PID"
+	}
+
+	podNames, err := framework.GetReadyNGFPodNames(k8sClient, ngfNamespace, releaseName, timeoutConfig.GetTimeout)
+	Expect(err).ToNot(HaveOccurred())
+	Expect(podNames).ToNot(BeEmpty())
+
+	restartCount, err := getContainerRestartCount(containerName, podNames[0])
+	Expect(err).ToNot(HaveOccurred())
+
+	job, err := runNodeDebuggerJob(podNames[0], jobScript)
+	Expect(err).ToNot(HaveOccurred())
+
+	Expect(waitForContainerRestart(podNames[0], containerName, restartCount)).ToNot(HaveOccurred())
+
+	// propagation policy is set to delete underlying pod created through job
+	Expect(resourceManager.Delete(
+		[]client.Object{job},
+		client.PropagationPolicy(metav1.DeletePropagationBackground),
```

Comment:
Looks like Background is default.
 https://kubernetes.io/docs/concepts/architecture/garbage-collection/#background-deletion 

What will happen if we don't specify metav1.DeletePropagationBackground ?

Reply:
https://kubernetes.io/docs/concepts/workloads/controllers/job/#:~:text=have%20a%20default%20deletion%20policy%20of%20orphanDependents%20causing%20Pods%20created%20by%20an%20unmanaged%20Job%20to%20be%20left%20around%20after%20that%20Job%20is%20fully%20deleted.

ahh, I did not have the extra DeleteOption, then it gave me a warning that the underlying pod would not be deleted after deleting the Job. So I looked into it and the document I linked above explains it.